Magento 2.4.1 and 2.3.6 Release Notes (#8036)
magento/devdocs: src/guides/v2.4/release-notes/open-source-2-4-1.md

#### Over 15 security enhancements that help close remote code execution (RCE) and cross-site scripting (XSS) vulnerabilities

> No confirmed attacks related to these issues have occurred to date. However, certain vulnerabilities can potentially be exploited to access customer information or take over administrator sessions. Most of these issues require that an attacker first obtains access to the Admin. As a result, we remind you to take all necessary steps to protect your Admin, including but not limited to these efforts: IP allowlisting, [two-factor authentication]({{page.baseurl}}/security/two-factor-authentication.html), use of a VPN, the use of a unique location rather than `/admin`, and good password hygiene. See[Security Updates Available for Magento](https://helpx.adobe.com/security/products/magento/apsb20-59.html) for a discussion of these fixed issues.

Reviewer: ```suggestion
No confirmed attacks related to these issues have occurred to date. However, certain vulnerabilities can potentially be exploited to access customer information or take over administrator sessions. Most of these issues require that an attacker first obtains access to the Admin. As a result, we remind you to take all necessary steps to protect your Admin, including but not limited to these efforts: IP allowlisting, [two-factor authentication]({{page.baseurl}}/security/two-factor-authentication.html), use of a VPN, the use of a unique location rather than `/admin`, and good password hygiene. See [Security Updates Available for Magento](https://helpx.adobe.com/security/products/magento/apsb20-59.html) for a discussion of these fixed issues.
```

*  Cart expiry settings are no longer re-set when an inventory or price update occurs. Previously, when a cart was set to expire in 24 hours, and an inventory update or price update occurred, the indexers populated the `updated_at table`, which re-set the expiry time.

<!--- MC-33313-->

*  Magento no longer throws a fatal error when an administrator assigns a customer who has an active shopping cart to a customer group.

<!-- ENGCOM-7834 -->

> *  Selecting a toolbar option from  product page **Developer tools**  >  **Network**  when running the Chrome browser no longer initializes `toolbar.js` twice. _Fix submitted by Paweł Tylek in pull request [28838](https://github.com/magento/magento2/pull/28838)_. [GitHub-25934](https://github.com/magento/magento2/issues/25934)

Author: changed in both files

<!--- ENGCOM-7420-->

*  Magento now adds an admin user’s ACL role ID to the product category tree cache ID. This will limit the category trees that an admin with limited scope can see as expected. _Fix submitted by quangdo-aligent in pull request [27429](https://github.com/magento/magento2/pull/27429)_. [GitHub-28306](https://github.com/magento/magento2/issues/28306)

### GraphQL

<!-- ENGCOM-7661 7662 7559 MC-35646 32345-->

*  Shoppers can select gift message options during checkout. GraphQL now covers gift message options for different types of cart items. See [`setGiftOptionsOnCart` mutation](https://devdocs.magento.com/guides/graphql/mutations/set-gift-options.html) and [`updateCartItems` mutation](https://devdocs.magento.com/guides/v2.4/graphql/mutations/update-cart-items.html). _Fix submitted by Oleh Usik in pull requests [28519](https://github.com/magento/magento2/pull/28105), [27956](https://github.com/magento/magento2/pull/27956), [28072](https://github.com/magento/magento2/pull/28072), [28072](https://github.com/magento/magento2/pull/28072). [GitHub-28519](https://github.com/magento/magento2/issues/28519)

<!-- ENGCOM-7508 -->

> *  Customers and guests can write product reviews. Customers can also retrieve their product review histories. See [Create a product review](https://devdocs.magento.com/guides/v2.4/graphql/mutations/create-product-review.html) and [productReviewRatingsMetadata query](https://devdocs.magento.com/guides/v2.4/graphql/queries/product-review-ratings-metadata.html for information on retrieving information about the reviews infrastructure. _Fix submitted by Eduard Chitoraga in pull request [27882](https://github.com/magento/magento2/pull/27882)_. [GitHub-28523](https://github.com/magento/magento2/issues/28523)

Reviewer: ```suggestion
*  Customers and guests can write product reviews. Customers can also retrieve their product review histories. See [Create a product review](https://devdocs.magento.com/guides/v2.4/graphql/mutations/create-product-review.html) and [productReviewRatingsMetadata query](https://devdocs.magento.com/guides/v2.4/graphql/queries/product-review-ratings-metadata.html) for information on retrieving information about the reviews infrastructure. _Fix submitted by Eduard Chitoraga in pull request [27882](https://github.com/magento/magento2/pull/27882)_. [GitHub-28523](https://github.com/magento/magento2/issues/28523)
```

<!-- ENGCOM-7968 -->

*  The [`addProductsToCart` mutation](https://devdocs.magento.com/guides/v2.4/graphql/mutations/add-products-to-cart.html) allows you to add any type of product to the active cart. We recommend using this mutation instead of single-purpose mutations such as `addSimpleProductsToCart`. _Fix submitted by Yaroslav Rogoza in pull request [27914](https://github.com/magento/magento2/pull/27914)_. [GitHub-28524](https://github.com/magento/magento2/issues/28524)

<!-- ENGCOM-7512 -->

*  GraphQL now uses only an authorization token to retrieve a type of user and its ID. Previously, GraphQL used an active customer's cookies to retrieve this information when the authorization token was empty. _Fix submitted by Alexander Taranovsky in pull request [27373](https://github.com/magento/magento2/pull/27373)_. [GitHub-28040](https://github.com/magento/magento2/issues/28040)

<!-- ENGCOM-7839 32949-->

*  The schema for the `createCustomer` mutation now matches validation requirements. _Fix submitted by Michał Derlatka in pull request [28888](https://github.com/magento/magento2/pull/28888)_. [GitHub-28570](https://github.com/magento/magento2/issues/28570)

<!-- ENGCOM-7750 -->

> *  `updateCustomer` no longer allows you to set an invalid `INT` value in the `gender` argument. _Fix submitted by Alexander Taranovsky in pull request [28487](https://github.com/magento/magento2/pull/28487)_. [GitHub-28481](https://github.com/magento/magento2/issues/28481)

Reviewer: ```suggestion
*  The `updateCustomer` mutation no longer allows you to set an invalid `INT` value in the `gender` argument. _Fix submitted by Alexander Taranovsky in pull request [28487](https://github.com/magento/magento2/pull/28487)_. [GitHub-28481](https://github.com/magento/magento2/issues/28481)
```

<!-- ENGCOM-7743 -->

*  Custom attribute aggregations now return store-specific option values. _Fix submitted by Dmitriy Gallyamov in pull request [28647](https://github.com/magento/magento2/pull/28647)_. [GitHub-28572](https://github.com/magento/magento2/issues/28572)

<!-- ENGCOM-7707 -->

*  Price intervals that are returned in a product aggregation can now be numeric only (no wildcards permitted). _Fix submitted by Dmitriy Gallyamov in pull request [28745](https://github.com/magento/magento2/pull/28745)_. [GitHub-28628](https://github.com/magento/magento2/issues/28628)

<!-- ENGCOM-7729 -->

*  The `categoryList` query now returns the correct response when using fragments. _Fix submitted by Ulzii in pull request [28710](https://github.com/magento/magento2/pull/28710)_. [GitHub-28584](https://github.com/magento/magento2/issues/28584)

<!--- MC-34187-->

> *  The new `availableStores` query returns a list of `StoreConfigs` for other stores available under the same website (based on the current store). It does not expose the list of websites. See [`availableStores` query](https://devdocs.magento.com/guides/v2.4/graphql/queries/available-stores.html). [GitHub-28569](https://github.com/magento/magento2/issues/28569)

Reviewer: ```suggestion
*  The new `availableStores` query returns a list of configuration attributes for multiple stores available under the same website (based on the current store). It does not expose the list of websites. See [`availableStores` query](https://devdocs.magento.com/guides/v2.4/graphql/queries/available-stores.html). _Fix submitted by Dmitriy Gallyamov in pull request [28794](https://github.com/magento/magento2/pull/28794)_. [GitHub-28569](https://github.com/magento/magento2/issues/28569)
```

<!-- ENGCOM-7810 MC-34187-->

> *  The `availableStores` query now returns codes only for the stores under the same website as the current store. _Fix submitted by Dmitriy Gallyamov in pull request [28794](https://github.com/magento/magento2/pull/28794)_. [GitHub-28569](https://github.com/magento/magento2/issues/28569)

Reviewer: This is the same as the previous item. Delete it.

Author: deleted

*  Magento no longer throws an error when an administrator changes the **Date** field during Admin product creation or save when the Admin locale is Chinese or Japanese. [GitHub-24696](https://github.com/magento/magento2/issues/24696)

<!--- ENGCOM-7829-->

> *  Magento now checks area (frontend or `adminhtml`) before rendering inline translation markup. Previously, making an API call to `/rest/V1/integration/admin/token` with bad credentials while inline translation was enabled results in the inclusion of inline translation markup around the error message in the API response. _Fix submitted by Zach Nanninga in pull request [28856](https://github.com/magento/magento2/pull/28856)_. [GitHub-28656](https://github.com/magento/magento2/issues/28656)

Reviewer: ```suggestion
*  Magento now checks area (frontend or `adminhtml`) before rendering inline translation markup. Previously, making an API call to `POST /V1/integration/admin/token` with bad credentials while inline translation was enabled results in the inclusion of inline translation markup around the error message in the API response. _Fix submitted by Zach Nanninga in pull request [28856](https://github.com/magento/magento2/pull/28856)_. [GitHub-28656](https://github.com/magento/magento2/issues/28656)
```

*  Search criteria filters now work as expected for product attributes that are used during the creation of a configurable product. [GitHub-29126](https://github.com/magento/magento2/issues/29126)

<!-- ENGCOM-7785 -->

*  Customer group ID is now validated for POST `/rest/V1/customers` requests. Previously, Magento did not display an error message when you used an invalid `group_id` in these requests. _Fix submitted by Vadim Malesh in pull request [28903](https://github.com/magento/magento2/pull/28903)_. [GitHub-28064](https://github.com/magento/magento2/issues/28064)

<!-- ENGCOM-7611 -->

*  PHP sessions are not created by default for anonymous REST calls. Previously, sessions were started even when unnecessary. _Fix submitted by Maciej Pawłowski in pull request [26032](https://github.com/magento/magento2/pull/26032)_. [GitHub-7213](https://github.com/magento/magento2/issues/7213)

<!-- ENGCOM-7612 -->

> *  Using the REST API to update order information no longer overwrites existing item information. Previously, values in `info_buyRequest` were changed. _Fix submitted by Vadim Malesh in pull request [28483](https://github.com/magento/magento2/pull/28483)_. [GitHub-22431](https://github.com/magento/magento2/issues/22431)

Reviewer: ```suggestion
*  Using various REST endpoints to update orders that contain custom option no longer deletes the custom option information. Previously, values in `info_buyRequest` were deleted. _Fix submitted by Vadim Malesh in pull request [28483](https://github.com/magento/magento2/pull/28483)_. [GitHub-22431](https://github.com/magento/magento2/issues/22431)
```

<!-- ENGCOM-7665 -->

*  Using GET `/rest/storeview2/V1/customers/me` now honors **Share Customer Accounts** setting (Admin **Stores** >  **Configuration**  >  **Customers**  >  **Customer Configuration**  > **Account Sharing Options**) when retrieving user information. _Fix submitted by Pavel Bystritsky in pull request [28213](https://github.com/magento/magento2/pull/28213)_. [GitHub-26089](https://github.com/magento/magento2/issues/26089)

<!-- ENGCOM-7563 -->

*  Address-related extension attributes can be set as data arrays. Previously, Magento threw an error when a POST request that contained  `extension_attributes` was sent to `estimate-shipping-methods`. This bug, which was introduced by a previous commit, affected any extension attribute set on `Magento\Quote\Api\Data\AddressInterface` that were not objects. _Fix submitted by Alexander Menk in pull request [27338](https://github.com/magento/magento2/pull/27338)_. [GitHub-26682](https://github.com/magento/magento2/issues/26682)

<!-- ENGCOM-7618 -->

> *  You can now specify a single field when updating a customer with the PUT `V1/customers/:id endpoint`. Previously, the endpoint required you to specify the customer email and other fields that are required to create a customer. _Fix submitted by Vadim Malesh in pull request [28332](https://github.com/magento/magento2/pull/28332)_. [GitHub-21237](https://github.com/magento/magento2/issues/21237)

Reviewer: ```suggestion
*  You can now specify a single field when updating a customer with the PUT `V1/customers/:id` endpoint. Previously, the endpoint required you to specify the customer email and other fields that are required to create a customer. _Fix submitted by Vadim Malesh in pull request [28332](https://github.com/magento/magento2/pull/28332)_. [GitHub-21237](https://github.com/magento/magento2/issues/21237)
```

**Issue**: Merchants cannot log in to dotdigital from the Admin in Safari when dotdigital is enabled. See the [It's impossible to login in the dotdigital via admin panel when dotdigital account is enabled](https://support.magento.com/hc/en-us/articles/360050092291) Knowledge Base article.<!--- BUNDLE-2704-->

**Issue**: Vertex address validation does not work during payment when the shopper selects a shipping address that differs from the billing address. The issue will be fixed in Magento 2.4.2. See the [Magento 2.4.1 known issue: Vertex Address Validation message does not disappear after updating address](https://support.magento.com/hc/en-us/articles/360050139631) Knowledge Base article.<!--- BUNDLE-2733-->

**Issue**: Magento displays an empty page when a merchant saves a dotdigital Page Builder form on the Safari browser. **Workaround**: Reload the page to apply your changes. See the [Magento 2.4.1 known issue: empty page after saving dotdigital Page Builder form](https://support.magento.com/hc/en-us/articles/360049819092) Knowledge Base article.

> **Issue**: Under certain conditions, Magento automatically signs in customers to previously used Amazon accounts rather prompting them to log in during checkout with Amazon Pay. See the [Magento 2.4.1 known issue: unable to change Amazon account in Google Chrome](https://support.magento.com/hc/en-us/articles/360049814152) Knowledge Base article.<!--- BUNDLE-2743—>

Reviewer: ```suggestion
**Issue**: Under certain conditions, Magento automatically signs in customers to previously used Amazon accounts rather prompting them to log in during checkout with Amazon Pay. See the [Magento 2.4.1 known issue: unable to change Amazon account in Google Chrome](https://support.magento.com/hc/en-us/articles/360049814152) Knowledge Base article.<!--- BUNDLE-2743-->
```
